File: Chapter 3 (Searching problems)/Uninformed Search/C++/bfs.cpp
#include <bits/stdc++.h>
using namespace std;

// Graph stored as adjacency matrix (2D vector).
// For unweighted graphs, use 1 to indicate an edge and -1 for no edge.
vector<int> bfs(const vector<vector<int>> &g, int s, int t)
{
    int n = (int)g.size();
    queue<int> q;
    vector<int> par(n, -1);
    vector<char> vis(n, 0);
    q.push(s);
    vis[s] = 1;
    while (!q.empty())
    {
        int u = q.front();
        q.pop();
        if (u == t)
        {
            vector<int> path;
            for (int x = u; x != -1; x = par[x])
                path.push_back(x);
            reverse(path.begin(), path.end());
            return path;
        }
        for (int v = 0; v < n; ++v)
        {
            if (g[u][v] != -1 && !vis[v])
            {
                vis[v] = 1;
                par[v] = u;
                q.push(v);
            }
        }
    }
    return {};
}

int main()
{
    // Node labels for printing
    vector<string> names = {"A", "B", "C", "D", "E", "F"};
    int n = (int)names.size();
    // Build adjacency matrix with -1 as no edge, 1 as edge
    vector<vector<int>> g(n, vector<int>(n, -1));
    auto add_edge = [&](int u, int v)
    { g[u][v] = 1; };
    add_edge(0, 1); // A->B
    add_edge(0, 2); // A->C
    add_edge(1, 3); // B->D
    add_edge(1, 4); // B->E
    add_edge(2, 5); // C->F
    add_edge(4, 5); // E->F

    auto path = bfs(g, 0, 5); // A to F
    for (size_t i = 0; i < path.size(); ++i)
    {
        if (i)
            cout << " ";
        cout << names[path[i]];
    }
    cout << "\n";
}

File: Chapter 3 (Searching problems)/Uninformed Search/Python/bfs.py
import collections

# Graph stored as adjacency matrix (2D list), -1 indicates no edge, 1 indicates edge
def bfs(graph, start, goal):
    n = len(graph)
    q = collections.deque([start])
    visited = [False]*n
    visited[start] = True
    parent = [-1]*n
    while q:
        u = q.popleft()
        if u == goal:
            path = []
            while u != -1:
                path.append(u)
                u = parent[u]
            return list(reversed(path))
        for v in range(n):
            if graph[u][v] != -1 and not visited[v]:
                visited[v] = True
                parent[v] = u
                q.append(v)
    return []

if __name__ == "__main__":
    names = ['A','B','C','D','E','F']
    n = len(names)
    g = [[-1]*n for _ in range(n)]
    def add(u,v): g[u][v] = 1
    add(0,1); add(0,2); add(1,3); add(1,4); add(2,5); add(4,5)
    path_idx = bfs(g, 0, 5)
    print([names[i] for i in path_idx])